controllers/etcdcluster_controller.go

```diff
@@ -86,51 +115,131 @@ func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
-		peerName := nextAvailablePeerName(cluster, peers.Items)
-		log.V(1).Info("Insufficient peers for replicas, adding new peer",
-			"current-peers", len(peers.Items),
-			"desired-peers", cluster.Spec.Replicas,
-			"peer", peerName)
-		peer := peerForCluster(cluster, peerName)
-		if err := r.Create(ctx, peer); err != nil {
-			log.Error(err, "Failed to create peer",
+	if members == nil {
+		peers := &etcdv1alpha1.EtcdPeerList{}
+		if err := r.List(ctx, peers, client.MatchingFields{clusterNameSpecField: cluster.Name}); err != nil {
+			log.Error(err, "unable to list peers")
+			return ctrl.Result{}, nil, err
+		}
+
+		if cluster.Spec.Replicas != nil && int32(len(peers.Items)) < *cluster.Spec.Replicas {
+			// Create more peers
+			peerName := nextAvailablePeerName(cluster, peers.Items)
+			log.V(1).Info("Insufficient peers for replicas, adding new peer",
+				"current-peers", len(peers.Items),
+				"desired-peers", cluster.Spec.Replicas,
 				"peer", peerName)
-			return ctrl.Result{}, err
+			peer := peerForCluster(cluster, peerName)
+			if err := r.Create(ctx, peer); err != nil {
+				log.Error(err, "Failed to create peer",
+					"peer", peerName)
+				return ctrl.Result{}, nil, err
+			}
+			return ctrl.Result{}, &event.PeerCreatedEvent{Object: cluster, PeerName: peer.Name}, nil
 		}
-		return ctrl.Result{}, nil
+	} else {
+		log.V(2).Info("Cluster communication established")
 	}
+	return ctrl.Result{RequeueAfter: time.Second * 10}, nil, nil
+}
 
-	return ctrl.Result{}, nil
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdpeers,verbs=get;list;watch;create
+// +kubebuilder:rbac:groups=core,resources=endpoints,verbs=get;list
+
+func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	log := r.Log.WithValues("cluster", req.NamespacedName)
+
+	cluster := &etcdv1alpha1.EtcdCluster{}
+	if err := r.Get(ctx, req.NamespacedName, cluster); err != nil {
+		log.Error(err, "unable to fetch EtcdCluster resource")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+		// special 'exit early' case.
+	}
+
+	// Attempt to dial the etcd cluster, recording the cluster response if we can
+	members, err := r.connectToEtcd(ctx, cluster)
+	if err != nil {
+		log.Error(err, "Unable to contact etcd cluster")
+	}
+
+	// Perform a reconcile, getting back the desired result, any errors, and a clusterEvent. This is an internal concept
+	// and is not the same as the Kubernetes event, although it is used to make one later.
+	result, clusterEvent, err := r.reconcile(ctx, req.NamespacedName, members, cluster)
+	if err != nil {
+		log.Error(err, "Failed to reconcile")
```

Comment:
is it correct to discard this error after logging it? iiuc, the controller will requeue the resource if it errors, could we miss that requeue by not returning the error?

Reply:
If the reconcile fails we go on to update the status. If that also were to fail which of the two errors do we give? Or should we early exit if the reconcile fails and not update status?

Reply:
on intuition, i'd expect the reconcile error to be more valuable than the status update error. maybe we should concat them and return that? but we do not currently return either error, which worries me we might skip a requeue

Reply:
Concatination makes sense to me. I'll do that and return an error if either errors.